lib/github/github/file.ex

```diff
@@ -0,0 +1,41 @@
+defmodule BorsNG.GitHub.File do
+  @moduledoc """
+  The structure of GitHub pull request file
+  """
+
+  @type tjson :: map
+  @type t :: %BorsNG.GitHub.File{
+          filename: bitstring
+        }
+  defstruct(filename: "")
+
+  @doc """
+  Convert from Poison-decoded JSON to a Pr struct.
+  """
+  @spec from_json!(tjson) :: [t]
+  def from_json!(json) do
+    {:ok, files} = from_json(json)
+    files
+  end
+
+  @doc """
+  Convert from Poison-decoded JSON to a Pr struct.
+  """
+  @spec from_json(tjson) :: {:ok, [t]} | {:error}
+  def from_json([
+        %{
+          "filename" => f
+        }
+      ]) do
+    {:ok,
+     [
+       %BorsNG.GitHub.File{
+         filename: f
+       }
+     ]}
+  end
```

Comment:
Okay, looking at the github client call, I think I got it.

You will receive a body like this:

```json
[
  {...},
  {...},
  ...
]
```

The code in this way will always fail because you're doing a pattern that says something like "Receive a list that has a single map as element and that map must have the key 'filename'".

To achieve the objective you want, you should probably map into the results:
```elixir
// [%{}, %{}, %{} ...]
body = Poison.decode!(raw_body)

Enum.map(body, fn el ->
  %BorsNG.Github.File{filename: Map.fetch!(el, "filename"}
end)
```

Also, on the subject of separation of concerns, I think that this module should only receive one "file" per time and the github client should the mapping.

Ie:

```elixir
// this file
def from_json(%{"filename" => filename}),
  do: %BorsNG.Github.File(filename: filename)

// Github client
files =
  body
  |> Poison.decode!()
  |> Enum.map(&Github.File.from_json/1)
{:ok, files}
```

Reply:
Note that the Github API _should_ always return the filename, this is an expectation that we have. IMO we should fail hard if this turns out not to be true

Reply:
Thanks ⚡️